File: desktop/main/semantic/engine.ts
import { promises as fs } from "node:fs";
import path from "node:path";
import { buildDeterministicEdges, buildNodes } from "./graph";
import { repairGraph, validateGraph } from "./integrity";
import { parseMarkdown } from "./markdown";
import { Block, ENGINE_VERSION, RebuildReport, SCHEMA_VERSION } from "./models";
import {
  buildFilesState,
  discoverMarkdownFiles,
  ensureSidecar,
  loadEdges,
  loadNodes,
  loadState,
  saveEdges,
  saveNodes,
  saveState
} from "./storage";

function cacheKey(filePath: string, fileHash: string): string {
  const safe = filePath.replace(/\//g, "__").replace(/\./g, "_");
  return `${safe}.${fileHash}.json`;
}

function serializeBlock(block: Block): Record<string, unknown> {
  return {
    id: block.id,
    type: block.type,
    file: block.file,
    heading_path: block.headingPath,
    title: block.title,
    content: block.content,
    content_hash: block.contentHash,
    structure_hash: block.structureHash,
    last_seen_timestamp: block.lastSeenTimestamp,
    start_line: block.startLine,
    end_line: block.endLine,
    link_targets: block.linkTargets,
    annotations: block.annotations.map((item) => ({ relation: item.relation, raw_target: item.rawTarget }))
  };
}

function deserializeBlock(payload: Record<string, unknown>): Block | null {
  try {
    const type = String(payload.type ?? "");
    const annotations = Array.isArray(payload.annotations)
      ? payload.annotations
          .filter((item) => item && typeof item === "object")
          .map((item) => {
            const value = item as Record<string, unknown>;
            return { relation: String(value.relation), rawTarget: String(value.raw_target) };
          })
      : [];
    return {
      id: String(payload.id ?? ""),
      type: type as Block["type"],
      file: String(payload.file ?? ""),
      headingPath: Array.isArray(payload.heading_path) ? payload.heading_path.map((part) => String(part)) : [],
      title: String(payload.title ?? ""),
      content: String(payload.content ?? ""),
      contentHash: String(payload.content_hash ?? ""),
      structureHash: String(payload.structure_hash ?? ""),
      lastSeenTimestamp: Number(payload.last_seen_timestamp ?? Date.now()),
      startLine: Number(payload.start_line ?? 1),
      endLine: Number(payload.end_line ?? 1),
      linkTargets: Array.isArray(payload.link_targets) ? payload.link_targets.map((item) => String(item)) : [],
      annotations: annotations.map((item) => ({ relation: item.relation as Block["annotations"][number]["relation"], rawTarget: item.rawTarget }))
    };
  } catch {
    return null;
  }
}

export class SemanticEngine {
  readonly repoRoot: string;
  private readonly sidecarPromise: Promise<string>;

  constructor(repoRoot: string) {
    this.repoRoot = repoRoot;
    this.sidecarPromise = ensureSidecar(repoRoot);
  }

  private async sidecar(): Promise<string> {
    return this.sidecarPromise;
  }

  private async cachePath(filePath: string, fileHash: string): Promise<string> {
    const sidecar = await this.sidecar();
    return path.join(sidecar, "cache", cacheKey(filePath, fileHash));
  }

  private async loadCachedBlocks(filePath: string, fileHash: string): Promise<Block[] | null> {
    const cacheFile = await this.cachePath(filePath, fileHash);
    try {
      const raw = await fs.readFile(cacheFile, "utf8");
      const payload = JSON.parse(raw) as Array<Record<string, unknown>>;
      if (!Array.isArray(payload)) {
        return null;
      }
      const blocks = payload.map((item) => deserializeBlock(item)).filter((item): item is Block => item !== null);
      return blocks.length > 0 ? blocks : null;
    } catch {
      return null;
    }
  }

  private async saveCachedBlocks(filePath: string, fileHash: string, blocks: readonly Block[]): Promise<void> {
    const cacheFile = await this.cachePath(filePath, fileHash);
    await fs.mkdir(path.dirname(cacheFile), { recursive: true });
    await fs.writeFile(cacheFile, `${JSON.stringify(blocks.map((block) => serializeBlock(block)), null, 2)}\n`, "utf8");

    const safePrefix = filePath.replace(/\//g, "__").replace(/\./g, "_");
    const cacheDir = path.dirname(cacheFile);
    for (const entry of await fs.readdir(cacheDir)) {
      if (!entry.startsWith(`${safePrefix}.`) || entry === path.basename(cacheFile)) {
        continue;
      }
      await fs.unlink(path.join(cacheDir, entry)).catch(() => {});
    }
  }

  async rebuild(options: { full?: boolean } = {}): Promise<RebuildReport> {
    let forceFull = Boolean(options.full);
    const warnings: string[] = [];
    const sidecar = await this.sidecar();

    let previousState = await loadState(sidecar);
    if (!previousState) {
      forceFull = true;
      warnings.push("state_missing_or_invalid:forcing_full_rebuild");
      previousState = {};
    }

    if (String(previousState.schema_version ?? "") && previousState.schema_version !== SCHEMA_VERSION) {
      forceFull = true;
      warnings.push("schema_version_mismatch:forcing_full_rebuild");
    }
    if (String(previousState.engine_version ?? "") && previousState.engine_version !== ENGINE_VERSION) {
      forceFull = true;
      warnings.push("engine_version_mismatch:forcing_full_rebuild");
    }

    const files = await discoverMarkdownFiles(this.repoRoot);
    if (files.length === 0) {
      warnings.push("no_docs_found:expected_markdown_under_docs");
    }
    const filesState = await buildFilesState(this.repoRoot, files);

    const previousFilesRaw = previousState.files;
    const previousFilesState =
      previousFilesRaw && typeof previousFilesRaw === "object" ? (previousFilesRaw as Record<string, { file_hash?: string }>) : {};

    const deletedFiles = Object.keys(previousFilesState)
      .filter((file) => !files.includes(file))
      .sort();

    const changedFiles: string[] = [];
    for (const filePath of files) {
      if (forceFull) {
        changedFiles.push(filePath);
        continue;
      }
      const current = filesState[filePath];
      const previous = previousFilesState[filePath];
      if (!previous || previous.file_hash !== current.file_hash) {
        changedFiles.push(filePath);
      }
    }

    const allBlocks: Block[] = [];
    for (const filePath of files) {
      const integrity = filesState[filePath];
      if (!forceFull && !changedFiles.includes(filePath)) {
        const cached = await this.loadCachedBlocks(filePath, integrity.file_hash);
        if (cached) {
          allBlocks.push(...cached);
          continue;
        }
        changedFiles.push(filePath);
        warnings.push(`cache_miss:${filePath}:reparsed`);
      }

      const content = await fs.readFile(path.join(this.repoRoot, filePath), "utf8");
      const blocks = parseMarkdown(filePath, content, Date.now());
      allBlocks.push(...blocks);
      await this.saveCachedBlocks(filePath, integrity.file_hash, blocks);
    }

    let nodes = buildNodes(allBlocks);
    let edges = buildDeterministicEdges(allBlocks);

    const errors = validateGraph(nodes, edges);
    if (errors.length > 0) {
      warnings.push(...errors.map((item) => `invariant_violation:${item}`));
      const repaired = repairGraph(nodes, edges);
      nodes = repaired.nodes;
      edges = repaired.edges;
      warnings.push(...repaired.warnings);
    }

    await saveNodes(sidecar, nodes);
    await saveEdges(sidecar, edges);
    await saveState(sidecar, {
      schema_version: SCHEMA_VERSION,
      engine_version: ENGINE_VERSION,
      files: filesState
    });

    return {
      processed_files: files.length,
      changed_files: [...new Set(changedFiles)].sort(),
      deleted_files: deletedFiles,
      nodes_written: nodes.length,
      edges_written: edges.length,
      warnings
    };
  }

  async check(): Promise<string[]> {
    const sidecar = await this.sidecar();
    const nodes = await loadNodes(sidecar);
    const edges = await loadEdges(sidecar);
    if (nodes.length === 0 && edges.length === 0) {
      return ["missing_or_empty_graph"];
    }
    return validateGraph(nodes, edges);
  }

  async loadGraph(): Promise<{
    state: Record<string, unknown> | null;
    nodes: Awaited<ReturnType<typeof loadNodes>>;
    edges: Awaited<ReturnType<typeof loadEdges>>;
  }> {
    const sidecar = await this.sidecar();
    const [state, nodes, edges] = await Promise.all([loadState(sidecar), loadNodes(sidecar), loadEdges(sidecar)]);
    return { state, nodes, edges };
  }
}

File: src/axonize/engine.py
from __future__ import annotations

import json
from pathlib import Path

from axonize import ENGINE_VERSION, SCHEMA_VERSION
from axonize.graph import build_deterministic_edges, build_nodes
from axonize.integrity import repair_graph, validate_graph
from axonize.markdown_ast import parse_markdown
from axonize.models import Block, BlockType, RebuildReport, RelationAnnotation, RelationType
from axonize.storage import (
    build_files_state,
    discover_markdown_files,
    ensure_sidecar,
    load_edges,
    load_nodes,
    load_state,
    save_edges,
    save_nodes,
    save_state,
)


def _cache_key(file_path: str, file_hash: str) -> str:
    safe = file_path.replace("/", "__").replace(".", "_")
    return f"{safe}.{file_hash}.json"


def _serialize_block(block: Block) -> dict:
    return {
        "id": block.id,
        "type": block.type.value,
        "file": block.file,
        "heading_path": list(block.heading_path),
        "title": block.title,
        "content": block.content,
        "content_hash": block.content_hash,
        "structure_hash": block.structure_hash,
        "last_seen_timestamp": block.last_seen_timestamp,
        "start_line": block.start_line,
        "end_line": block.end_line,
        "link_targets": block.link_targets,
        "annotations": [{"relation": item.relation.value, "raw_target": item.raw_target} for item in block.annotations],
    }


def _deserialize_block(payload: dict) -> Block | None:
    try:
        annotations = [
            RelationAnnotation(RelationType(item["relation"]), str(item["raw_target"]))
            for item in payload.get("annotations", [])
        ]
        return Block(
            id=str(payload["id"]),
            type=BlockType(str(payload["type"])),
            file=str(payload["file"]),
            heading_path=tuple(str(part) for part in payload.get("heading_path", [])),
            title=str(payload.get("title", "")),
            content=str(payload.get("content", "")),
            content_hash=str(payload.get("content_hash", "")),
            structure_hash=str(payload.get("structure_hash", "")),
            last_seen_timestamp=int(payload.get("last_seen_timestamp", 0)),
            start_line=int(payload.get("start_line", 1)),
            end_line=int(payload.get("end_line", 1)),
            link_targets=[str(item) for item in payload.get("link_targets", [])],
            annotations=annotations,
        )
    except (KeyError, TypeError, ValueError):
        return None


class SemanticEngine:
    def __init__(self, repo_root: Path) -> None:
        self.repo_root = repo_root
        self.sidecar = ensure_sidecar(repo_root)

    def _cache_file_path(self, file_path: str, file_hash: str) -> Path:
        return self.sidecar / "cache" / _cache_key(file_path, file_hash)

    def _load_cached_blocks(self, file_path: str, file_hash: str) -> list[Block] | None:
        cache_path = self._cache_file_path(file_path, file_hash)
        if not cache_path.exists():
            return None
        try:
            payload = json.loads(cache_path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError):
            return None
        if not isinstance(payload, list):
            return None

        blocks: list[Block] = []
        for item in payload:
            if not isinstance(item, dict):
                continue
            block = _deserialize_block(item)
            if block is None:
                continue
            blocks.append(block)
        return blocks or None

    def _save_cached_blocks(self, file_path: str, file_hash: str, blocks: list[Block]) -> None:
        cache_path = self._cache_file_path(file_path, file_hash)
        cache_path.parent.mkdir(parents=True, exist_ok=True)
        cache_path.write_text(
            json.dumps([_serialize_block(block) for block in blocks], ensure_ascii=True, indent=2, sort_keys=True) + "\n",
            encoding="utf-8",
        )

        safe_prefix = file_path.replace("/", "__").replace(".", "_")
        for stale in cache_path.parent.glob(f"{safe_prefix}.*.json"):
            if stale.name != cache_path.name:
                stale.unlink(missing_ok=True)

    def rebuild(self, full: bool = False) -> RebuildReport:
        warnings: list[str] = []
        previous_state = load_state(self.sidecar)

        if previous_state is None:
            full = True
            warnings.append("state_missing_or_invalid:forcing_full_rebuild")
            previous_state = {}

        prev_schema = previous_state.get("schema_version")
        prev_engine = previous_state.get("engine_version")
        if prev_schema and prev_schema != SCHEMA_VERSION:
            full = True
            warnings.append("schema_version_mismatch:forcing_full_rebuild")
        if prev_engine and prev_engine != ENGINE_VERSION:
            full = True
            warnings.append("engine_version_mismatch:forcing_full_rebuild")

        files = discover_markdown_files(self.repo_root)
        files_state = build_files_state(self.repo_root, files)

        prev_files_state = previous_state.get("files", {})
        if not isinstance(prev_files_state, dict):
            prev_files_state = {}

        deleted_files = sorted(set(prev_files_state.keys()) - set(files))
        changed_files: list[str] = []
        for file_path in files:
            if full:
                changed_files.append(file_path)
                continue
            current = files_state[file_path]
            previous = prev_files_state.get(file_path)
            if not isinstance(previous, dict):
                changed_files.append(file_path)
                continue
            if previous.get("file_hash") != current.file_hash:
                changed_files.append(file_path)

        all_blocks: list[Block] = []
        for file_path in files:
            file_integrity = files_state[file_path]
            if not full and file_path not in changed_files:
                cached = self._load_cached_blocks(file_path, file_integrity.file_hash)
                if cached is not None:
                    all_blocks.extend(cached)
                    continue
                changed_files.append(file_path)
                warnings.append(f"cache_miss:{file_path}:reparsed")

            content = (self.repo_root / file_path).read_text(encoding="utf-8")
            blocks = parse_markdown(file_path, content)
            all_blocks.extend(blocks)
            self._save_cached_blocks(file_path, file_integrity.file_hash, blocks)

        nodes = build_nodes(all_blocks)
        edges = build_deterministic_edges(all_blocks)

        errors = validate_graph(nodes, edges)
        if errors:
            warnings.extend([f"invariant_violation:{item}" for item in errors])
            nodes, edges, repair_warnings = repair_graph(nodes, edges)
            warnings.extend(repair_warnings)

        save_nodes(self.sidecar, nodes)
        save_edges(self.sidecar, edges)

        state_payload = {
            "schema_version": SCHEMA_VERSION,
            "engine_version": ENGINE_VERSION,
            "files": {file: state.to_dict() for file, state in files_state.items()},
        }
        save_state(self.sidecar, state_payload)

        return RebuildReport(
            processed_files=len(files),
            changed_files=sorted(changed_files),
            deleted_files=deleted_files,
            nodes_written=len(nodes),
            edges_written=len(edges),
            warnings=warnings,
        )

    def check(self) -> list[str]:
        nodes = load_nodes(self.sidecar)
        edges = load_edges(self.sidecar)
        if not (self.sidecar / "nodes.json").exists():
            return ["missing_nodes_json"]

        return validate_graph(nodes, edges)
